### PR TITLE
Batched version implementation

### PR DESCRIPTION
**Pull request by SonicFrog**
This implements the batched, latency aware optimized version of murmur

### CODE REVIEW

File: src/batched/mod.rs [hunk 1]
```diff
@@ -1,18 +1,17 @@
-//! This module provides an optimized implementation of [`Murmur`] using batching and bittorrent-like distribution of
+//! <p> This module provides an optimized implementation of [`Murmur`] using batching and bittorrent-like distribution of
 //! [`Batch`]es. Batch creation involves payload collecting node (rendezvous nodes) that are reponsible for collating all
-//! payloads into a [`Batch`] before disseminating the [`Batch`] throughout the network.
+//! payloads into a [`Batch`] before disseminating the [`Batch`] throughout the network.</p>
 //!
-//! Different rendez vous policy are supported using the [`RdvPolicy`] trait.
+//! <p>Different rendez vous policy are supported using the [`RdvPolicy`] trait.</p>
```
Comment (tharvik) [inline src/batched/mod.rs:5]: this block doesn't appear in the generated doc

File: src/batched/mod.rs [hunk 2]
```diff
@@ -239,69 +261,56 @@ where
...
-        OptionFuture::from(self.batches.read().await.get(batch).map(|batch| {
-            batch.check().then(move |need| async move {
-                if need.map_or_else(|e| e.valid(), |_| true) {
-                    batch.complete().map_ok_or_else(|_| false, |_| true).await
+    async fn try_deliver(&self, digest: &Digest) -> Result<Option<Arc<Batch<M>>>, MurmurError> {
+        let mut guard = self.batches.write().await;
+
+        match guard.entry(*digest) {
+            Entry::Vacant(_) => Ok(None),
+            Entry::Occupied(mut e) => {
```
Comment (tharvik) [inline src/batched/mod.rs:269]: you can simplify it to `Ok(e.get_mut().ready().await)`

File: examples/node/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[package]
```
[the remark this answers is not in the document]
Comment (SonicFrog) [reply examples/node/Cargo.toml:1]: This example never returns if no other nodes are running

File: src/batched/config.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    /// Get the sponge threshold
+    pub fn sponge_threshold(&self) -> usize {
+        self.sponge_threshold
+    }
+
+    /// Get the block size
+    pub fn block_size(&self) -> usize {
+        self.block_size
+    }
```
[the remark this answers is not in the document]
Comment (SonicFrog) [reply src/batched/config.rs:66]: Duration does not implement FromStr which means you can't get one directly from the command line arguments

File: examples/node/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tharvik) [reply examples/node/Cargo.toml:1]: no need to `cargo run` it, simply to build it to ensure that it is in sync with the main crate

File: .projectile [hunk 5]
```diff
@@ -0,0 +1 @@
+/target/
```
[the remark this answers is not in the document]
Comment (SonicFrog) [reply .projectile:1]: Yep

[the remark this answers is not in the document]
Comment (tharvik) [reply .projectile:1]: you can use `$HOME/.config/git/ignore` to not track some user-specific files (and there is not smiley for nitpicking, too bad)

File: src/batched/sync.rs [hunk 6]
```diff
@@ -313,73 +315,4 @@ pub mod test {
             assert_eq!(actual, expected);
         }
     }
-
-    #[tokio::test]
```
[the remark this answers is not in the document]
Comment (SonicFrog) [reply src/batched/sync.rs:317]: These tests were for features that are not there anymore post refactoring

File: examples/node/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tharvik) [reply examples/node/Cargo.toml:1]: too bad, forget it then

File: src/batched/sync.rs [hunk 7]
```diff
@@ -237,6 +152,95 @@ impl<M: Message + 'static> BatchState<M> {
...
+        write!(
+            f,
+            "{} with {} blocks",
+            self.info.digest(),
+            self.info.sequence()
+        )
+    }
+}
+
+/// Temporary wrapper until tokio allows mapping RwLockWriteGuard again
```
[the remark this answers is not in the document]
Comment (SonicFrog) [reply src/batched/sync.rs]: Newest version of tokio actually re-added mapping in RwLockWriteGuard, I'll switch it back

**Review by tharvik — APPROVED** [review]
Yeay, first PR accepted :1st_place_medal:

**Pull request merged**